// vue-language-server/dist/modes/style/stylus/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const _ = require("lodash");
const emmet = require("vscode-emmet-helper");
const emmet_1 = require("../emmet");
const languageModelCache_1 = require("../../languageModelCache");
const completion_item_1 = require("./completion-item");
const symbols_finder_1 = require("./symbols-finder");
const stylus_hover_1 = require("./stylus-hover");
function getStylusMode(documentRegions) {
    const embeddedDocuments = languageModelCache_1.getLanguageModelCache(10, 60, document => documentRegions.get(document).getEmbeddedDocument('stylus'));
    return {
        getId: () => 'stylus',
        onDocumentRemoved() { },
        dispose() { },
        doComplete(document, position) {
            const embedded = embeddedDocuments.get(document);
            const emmetCompletions = emmet.doComplete(document, position, 'stylus', {
                useNewEmmet: true,
                showExpandedAbbreviation: true,
                showAbbreviationSuggestions: true,
                syntaxProfiles: {},
                variables: {}
            });
            const emmetItems = _.map(emmetCompletions.items, i => {
                return Object.assign({}, i, { sortText: emmet_1.Priority.Emmet + i.label });
            });
            const lsCompletions = completion_item_1.provideCompletionItems(embedded, position);
            const lsItems = _.map(lsCompletions.items, i => {
                return Object.assign({}, i, { sortText: emmet_1.Priority.Platform + i.label });
            });
            return {
                isIncomplete: true,
                items: _.concat(emmetItems, lsItems)
            };
        },
        findDocumentSymbols(document) {
            const embedded = embeddedDocuments.get(document);
            return symbols_finder_1.provideDocumentSymbols(embedded);
        },
        doHover(document, position) {
            const embedded = embeddedDocuments.get(document);
            return stylus_hover_1.stylusHover(embedded, position);
        },
    };
}
exports.getStylusMode = getStylusMode;
exports.wordPattern = /(#?-?\d*\.\d\w*%?)|([$@#!.:]?[\w-?]+%?)|[$@#!.]/g;
