// atom-languageclient/build/test/adapters/linter-push-v2-adapter.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const linter_push_v2_adapter_1 = require("../../lib/adapters/linter-push-v2-adapter");
const convert_1 = require("../../lib/convert");
const ls = require("../../lib/languageclient");
const path = require("path");
const sinon = require("sinon");
const chai_1 = require("chai");
const atom_1 = require("atom");
const helpers_js_1 = require("../helpers.js");
describe('LinterPushV2Adapter', () => {
    beforeEach(() => {
        global.sinon = sinon.sandbox.create();
    });
    afterEach(() => {
        global.sinon.restore();
    });
    describe('constructor', () => {
        it('subscribes to onPublishDiagnostics', () => {
            const languageClient = new ls.LanguageClientConnection(helpers_js_1.createSpyConnection());
            sinon.spy(languageClient, 'onPublishDiagnostics');
            new linter_push_v2_adapter_1.default(languageClient);
            chai_1.expect(languageClient.onPublishDiagnostics.called).equals(true);
        });
    });
    describe('diagnosticToMessage', () => {
        it('converts Diagnostic and path to a linter$Message', () => {
            const filePath = '/a/b/c/d';
            const diagnostic = {
                message: 'This is a message',
                range: {
                    start: { line: 1, character: 2 },
                    end: { line: 3, character: 4 },
                },
                source: 'source',
                code: 'code',
                severity: ls.DiagnosticSeverity.Information,
            };
            const connection = { onPublishDiagnostics() { } };
            const adapter = new linter_push_v2_adapter_1.default(connection);
            const result = adapter.diagnosticToV2Message(filePath, diagnostic);
            chai_1.expect(result.excerpt).equals(diagnostic.message);
            chai_1.expect(result.linterName).equals(diagnostic.source);
            chai_1.expect(result.location.file).equals(filePath);
            chai_1.expect(result.location.position).deep.equals(new atom_1.Range(new atom_1.Point(1, 2), new atom_1.Point(3, 4)));
            chai_1.expect(result.severity).equals('info');
        });
    });
    describe('diagnosticSeverityToSeverity', () => {
        it('converts DiagnosticSeverity.Error to "error"', () => {
            const severity = linter_push_v2_adapter_1.default.diagnosticSeverityToSeverity(ls.DiagnosticSeverity.Error);
            chai_1.expect(severity).equals('error');
        });
        it('converts DiagnosticSeverity.Warning to "warning"', () => {
            const severity = linter_push_v2_adapter_1.default.diagnosticSeverityToSeverity(ls.DiagnosticSeverity.Warning);
            chai_1.expect(severity).equals('warning');
        });
        it('converts DiagnosticSeverity.Information to "info"', () => {
            const severity = linter_push_v2_adapter_1.default.diagnosticSeverityToSeverity(ls.DiagnosticSeverity.Information);
            chai_1.expect(severity).equals('info');
        });
        it('converts DiagnosticSeverity.Hint to "info"', () => {
            const severity = linter_push_v2_adapter_1.default.diagnosticSeverityToSeverity(ls.DiagnosticSeverity.Hint);
            chai_1.expect(severity).equals('info');
        });
    });
    describe('captureDiagnostics', () => {
        it('stores diagnostic codes and allows their retrival', () => {
            const languageClient = new ls.LanguageClientConnection(helpers_js_1.createSpyConnection());
            const adapter = new linter_push_v2_adapter_1.default(languageClient);
            const testPath = path.join(__dirname, 'test.txt');
            adapter.captureDiagnostics({
                uri: convert_1.default.pathToUri(testPath),
                diagnostics: [
                    {
                        message: 'Test message',
                        range: {
                            start: { line: 1, character: 2 },
                            end: { line: 3, character: 4 },
                        },
                        source: 'source',
                        code: 'test code',
                        severity: ls.DiagnosticSeverity.Information,
                    },
                ],
            });
            const mockEditor = helpers_js_1.createFakeEditor(testPath);
            chai_1.expect(adapter.getDiagnosticCode(mockEditor, new atom_1.Range([1, 2], [3, 4]), 'Test message')).to.equal('test code');
            chai_1.expect(adapter.getDiagnosticCode(mockEditor, new atom_1.Range([1, 2], [3, 4]), 'Test message2')).to.not.exist;
            chai_1.expect(adapter.getDiagnosticCode(mockEditor, new atom_1.Range([1, 2], [3, 5]), 'Test message')).to.not.exist;
        });
    });
});
