// omnisharp-client/lib/helpers/platform.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var SupportedPlatform;
(function (SupportedPlatform) {
    SupportedPlatform[SupportedPlatform["None"] = 0] = "None";
    SupportedPlatform[SupportedPlatform["Windows"] = 1] = "Windows";
    SupportedPlatform[SupportedPlatform["OSX"] = 2] = "OSX";
    SupportedPlatform[SupportedPlatform["Linux"] = 3] = "Linux";
})(SupportedPlatform = exports.SupportedPlatform || (exports.SupportedPlatform = {}));
/* tslint:disable:quotemark no-unexternalized-strings */
function getSupportedPlatform(platform) {
    if (platform === void 0) { platform = process.platform; }
    if (process.platform === 'win32') {
        return SupportedPlatform.Windows;
    }
    else if (process.platform === 'darwin') {
        return SupportedPlatform.OSX;
    }
    else if (process.platform === 'linux') {
        return SupportedPlatform.Linux;
    }
    return SupportedPlatform.None;
}
exports.getSupportedPlatform = getSupportedPlatform;
/* tslint:enable:quotemark */
exports.supportedPlatform = getSupportedPlatform(process.platform);
