// atom-haskell-utils/lib/util.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const atom_1 = require("atom");
const fs = require("fs");
const path = require("path");
function hasGetPath(dir) {
    return dir && dir.getPath && typeof dir.getPath === 'function';
}
function isTextBuffer(x) {
    return x && x.file;
}
function isDirectory(dir) {
    if (dir === null) {
        return false;
    }
    if (hasGetPath(dir)) {
        return isDirectory(dir.getPath());
    }
    else if (typeof dir === 'string') {
        try {
            return fs.statSync(dir).isDirectory();
        }
        catch (e) {
            return false;
        }
    }
    else {
        return false;
    }
}
exports.isDirectory = isDirectory;
function getRootDirFallback(file) {
    let dir = null;
    if (file) {
        ;
        [dir] = atom.project
            .getDirectories()
            .filter((d) => d.contains(file.getPath()));
    }
    if (!dir) {
        dir = atom.project.getDirectories()[0];
    }
    if (dir && dir.getPath() === 'atom://config') {
        dir = null;
    }
    if (!(dir && isDirectory(dir))) {
        if (file) {
            dir = file.getParent();
        }
        else {
            dir = new atom_1.Directory('.');
        }
    }
    return dir;
}
exports.getRootDirFallback = getRootDirFallback;
async function getDirEntries(dir) {
    return new Promise((resolve, reject) => dir.getEntries((error, contents) => {
        if (error) {
            reject(error);
        }
        else {
            resolve(contents);
        }
    }));
}
exports.getDirEntries = getDirEntries;
async function dirHasCabalFile(d) {
    if (!d) {
        return false;
    }
    return (await getDirEntries(d)).some((file) => file.isFile() && file.getBaseName().endsWith('.cabal'));
}
async function dirHasSandboxFile(d) {
    if (!d) {
        return false;
    }
    return (await getDirEntries(d)).some((file) => file.isFile() && file.getBaseName() === 'cabal.sandbox.config');
}
async function findProjectRoot(d, check) {
    while (!(d.isRoot() || (await check(d)))) {
        d = d && d.getParent();
    }
    if (await check(d)) {
        return d;
    }
    else {
        return null;
    }
}
async function getRootDir(input) {
    let file;
    if (isTextBuffer(input)) {
        const p = input.getPath();
        if (p)
            file = new atom_1.File(p);
        else
            file = null;
    }
    else if (hasGetPath(input)) {
        file = input;
    }
    else if (typeof input === 'string') {
        file = new atom_1.File(input);
    }
    else {
        file = null;
    }
    let dir;
    if (file && isDirectory(file)) {
        dir = new atom_1.Directory(file.getPath());
    }
    else {
        dir = (file && file.getParent()) || getRootDirFallback(file);
    }
    const cabalRoot = await findProjectRoot(dir, dirHasCabalFile);
    const sandboxRoot = await findProjectRoot(dir, dirHasSandboxFile);
    dir = cabalRoot || sandboxRoot;
    if (!(dir && isDirectory(dir))) {
        dir = getRootDirFallback(file);
    }
    return dir;
}
exports.getRootDir = getRootDir;
const HS = require("../hs/hs.min.js");
async function parseDotCabal(cabalSource) {
    return new Promise((resolve) => {
        HS.parseDotCabal(cabalSource, resolve);
    });
}
exports.parseDotCabal = parseDotCabal;
async function getComponentFromFile(cabalSource, filePath) {
    const fp = process.platform === 'win32'
        ? filePath.replace(path.sep, path.posix.sep)
        : filePath;
    return new Promise((resolve) => {
        HS.getComponentFromFile(cabalSource, fp, resolve);
    });
}
exports.getComponentFromFile = getComponentFromFile;
async function unlit(filename, source) {
    return new Promise((resolve, reject) => {
        HS.unlit(filename, source, (error, result) => {
            if (error) {
                reject(new Error(error));
            }
            else if (result) {
                resolve(result);
            }
            else {
                reject(new Error('Unknown error when trying to run unlit'));
            }
        });
    });
}
exports.unlit = unlit;
function isErrorResult(x) {
    return x && x.error && typeof x.error === 'string';
}
async function parseHsModuleImports(source) {
    return new Promise((resolve, reject) => {
        HS.parseHsModuleImports(source, (result) => {
            if (isErrorResult(result)) {
                reject(new Error(result.error));
            }
            else {
                resolve(result);
            }
        });
    });
}
exports.parseHsModuleImports = parseHsModuleImports;
exports.hsEscapeString = HS.hsEscapeString;
