// yaml-language-server/out/server/src/languageService/yamlLanguageService.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Red Hat, Inc. All rights reserved.
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
const jsonSchemaService_1 = require("./services/jsonSchemaService");
const documentSymbols_1 = require("./services/documentSymbols");
const yamlCompletion_1 = require("./services/yamlCompletion");
const yamlHover_1 = require("./services/yamlHover");
const yamlValidation_1 = require("./services/yamlValidation");
function getLanguageService(schemaRequestService, workspaceContext, contributions, customSchemaProvider, promiseConstructor) {
    let promise = promiseConstructor || Promise;
    let schemaService = new jsonSchemaService_1.JSONSchemaService(schemaRequestService, workspaceContext, customSchemaProvider);
    let completer = new yamlCompletion_1.YAMLCompletion(schemaService, contributions, promise);
    let hover = new yamlHover_1.YAMLHover(schemaService, contributions, promise);
    let yamlDocumentSymbols = new documentSymbols_1.YAMLDocumentSymbols();
    let yamlValidation = new yamlValidation_1.YAMLValidation(schemaService, promise);
    return {
        configure: (settings) => {
            schemaService.clearExternalSchemas();
            if (settings.schemas) {
                settings.schemas.forEach(settings => {
                    schemaService.registerExternalSchema(settings.uri, settings.fileMatch, settings.schema);
                });
            }
            yamlValidation.configure(settings);
        },
        doComplete: completer.doComplete.bind(completer),
        doResolve: completer.doResolve.bind(completer),
        doValidation: yamlValidation.doValidation.bind(yamlValidation),
        doHover: hover.doHover.bind(hover),
        findDocumentSymbols: yamlDocumentSymbols.findDocumentSymbols.bind(yamlDocumentSymbols),
        resetSchema: (uri) => schemaService.onResourceChange(uri)
    };
}
exports.getLanguageService = getLanguageService;
