// flow-language-server/lib/DocumentHighlight.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _asyncToGenerator = require('async-to-generator');

var _asyncToGenerator2 = _interopRequireDefault(_asyncToGenerator);

var _FlowSingleProjectLanguageService = require('./pkg/nuclide-flow-rpc/lib/FlowSingleProjectLanguageService');

var _TextDocuments = require('./TextDocuments');

var _TextDocuments2 = _interopRequireDefault(_TextDocuments);

var _util = require('./utils/util');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

class DocumentHighlightSupport {

  constructor({ documents, flow }) {
    this.documents = documents;
    this.flow = flow;
  }

  provideDocumentHighlight(params) {
    var _this = this;

    return (0, _asyncToGenerator2.default)(function* () {
      const position = params.position,
            textDocument = params.textDocument;


      const fileName = (0, _util.fileURIToPath)(textDocument.uri);
      const doc = _this.documents.get(textDocument.uri);

      const highlights = yield _this.flow.highlight(fileName, doc.buffer, (0, _util.lspPositionToAtomPoint)(position));
      if (highlights) {
        return highlights.map(function (highlight) {
          return {
            range: {
              start: (0, _util.atomPointToLSPPosition)(highlight.start),
              end: (0, _util.atomPointToLSPPosition)(highlight.end)
            }
          };
        });
      }
      return null;
    })();
  }
}
exports.default = DocumentHighlightSupport; /**
                                             * Copyright (c) 2017-present, Facebook, Inc.
                                             * All rights reserved.
                                             *
                                             * This source code is licensed under the BSD-style license found in the
                                             * LICENSE file in the root directory of this source tree. An additional grant
                                             * of patent rights can be found in the PATENTS file in the same directory.
                                             *
                                             * 
                                             * @format
                                             */
