// vue-language-server/dist/modes/template/tagProviders/onsenTags.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const common_1 = require("./common");
const tags = require("vue-onsenui-helper-json/vue-onsenui-tags.json");
const attributes = require("vue-onsenui-helper-json/vue-onsenui-attributes.json");
function getOnsenTagProvider() {
    return {
        getId: () => 'onsen',
        priority: common_1.Priority.Library,
        collectTags(collector) {
            for (const tagName in tags) {
                collector(tagName, tags[tagName].description || '');
            }
        },
        collectAttributes(tag, collector) {
            if (!tags[tag]) {
                return;
            }
            const attrs = tags[tag].attributes;
            if (!attrs) {
                return;
            }
            for (const attr of attrs) {
                const detail = findAttributeDetail(tag, attr);
                collector(attr, undefined, detail && detail.description || '');
            }
        },
        collectValues(tag, attr, collector) {
            if (!tags[tag]) {
                return;
            }
            const attrs = tags[tag].attributes;
            if (!attrs || attrs.indexOf(attr) < 0) {
                return;
            }
            const detail = findAttributeDetail(tag, attr);
            if (!detail || !detail.options) {
                return;
            }
            for (const option of detail.options) {
                collector(option);
            }
        }
    };
}
exports.getOnsenTagProvider = getOnsenTagProvider;
function findAttributeDetail(tag, attr) {
    return attributes[attr] || attributes[tag + '/' + attr];
}
